src/main/java/org/jahia/modules/saml2/SAML2Util.java

```diff
@@ -111,17 +109,34 @@ public SAML2ClientConfiguration getSAML2ClientConfiguration(SAML2Settings saml2S
      * @param request
      */
     private SAML2Client initSAMLClient(SAML2Settings saml2Settings, HttpServletRequest request) {
-        String spMetaDataLocation = getSamlFileName(saml2Settings.getSiteKey(), "sp-metadata.xml");
-        final File spMetadataFile = new File(spMetaDataLocation);
-        if (spMetadataFile.exists()) {
-            spMetadataFile.delete();
+        return initSAMLClient(getSAML2ClientConfiguration(saml2Settings), getAssertionConsumerServiceUrl(request, saml2Settings.getIncomingTargetUrl()));
+    }
+
+    private SAML2Client initSAMLClient(SAML2ClientConfiguration saml2ClientConfiguration, String callbackUrl) {
+        try {
+            final File spMetadataFile = saml2ClientConfiguration.getServiceProviderMetadataResource().getFile();
+            if (spMetadataFile.exists()) {
+                spMetadataFile.delete();
+            }
+        } catch (IOException e) {
+            throw new TechnicalException("Cannot udpate SP Metadata file", e);
         }
 
-        final SAML2Client client = new SAML2Client(getSAML2ClientConfiguration(saml2Settings));
-        client.setCallbackUrl(getAssertionConsumerServiceUrl(request, saml2Settings.getIncomingTargetUrl()));
+        final SAML2Client client = new SAML2Client(saml2ClientConfiguration);
+        client.setCallbackUrl(callbackUrl);
+        try {
+            client.init();
+        } catch (NullPointerException e) {
+            if (e.getStackTrace().length > 0 && e.getStackTrace()[0].getClassName().equals("org.opensaml.saml.metadata.resolver.impl.DOMMetadataResolver")) {
```

Comment:
Do you by any chance remember what is the origin of this NPE? May worth it to provide some comment here to either be able to track a bug in this third-party library or just have some history if this "hack" become insufficient or irrelevant in a distant future.

Reply:
Yes, I did not find a better way, but basically if you provide an XML which is not a saml metadata, the DOMMetadataResolver fails with NPE at line 68 because it cannot find an "unmarshaller" for an unknown xml element. This is just to avoid sending an "unknown error" message
```suggestion
            // Check if we have an NPE in DOMMetadataResolver, meaning we get an unknown XML element
            if (e.getStackTrace().length > 0 && e.getStackTrace()[0].getClassName().equals("org.opensaml.saml.metadata.resolver.impl.DOMMetadataResolver")) {
```